bofire/models/diagnostics.py

```diff
@@ -0,0 +1,180 @@
+from typing import Optional, Sequence
+
+import numpy as np
+import pandas as pd
+from pydantic import root_validator, validator
+from sklearn.metrics import (
+    mean_absolute_error,
+    mean_absolute_percentage_error,
+    mean_squared_error,
+    r2_score,
+)
+
+from bofire.domain.util import PydanticBaseModel, is_numeric
+from bofire.utils.enum import RegressionMetricsEnum
+
+metrics = {
+    RegressionMetricsEnum.MAE: mean_absolute_error,
+    RegressionMetricsEnum.MSD: mean_squared_error,
+    RegressionMetricsEnum.R2: r2_score,
+    RegressionMetricsEnum.MAPE: mean_absolute_percentage_error,
+}
+
+
+class CVResult(PydanticBaseModel):
+    """Container representing the results of one CV fold.
+
+    Attributes:
+        key (str): Key of the validated output feature.
+        observed (pd.Series): Series holding the observed values
+        predicted (pd.Series): Series holding the predicted values
+        uncertainty (pd.Series, optional): Series holding the uncertainty associated with
+            the prediction. Defaults to None.
+    """
+
+    key: str
+    observed: pd.Series
+    predicted: pd.Series
+    uncertainty: Optional[pd.Series] = None
+
+    @root_validator(pre=True)
+    def validate_shapes(cls, values):
+        if not len(values["predicted"]) == len(values["observed"]):
+            raise ValueError(
+                f"Predicted values has length {len(values['predicted'])} whereas observed has length {len(values['observed'])}"
+            )
+        if "uncertainty" in values:
+            if not len(values["predicted"]) == len(values["uncertainty"]):
+                raise ValueError(
+                    f"Predicted values has length {len(values['predicted'])} whereas uncertainty has length {len(values['uncertainty'])}"
+                )
+        return values
+
+    @validator("observed")
+    def validate_observed(cls, v, values):
+        if not is_numeric(v):
+            raise ValueError("Not all values of observed are numerical")
+        return v
+
+    @validator("predicted")
+    def validate_predicted(cls, v, values):
+        if not is_numeric(v):
+            raise ValueError("Not all values of observed are numerical")
+        return v
+
+    @validator("uncertainty")
+    def validate_uncertainty(cls, v, values):
+        if not is_numeric(v):
+            raise ValueError("Not all values of observed are numerical")
+        return v
+
+    @property
+    def num_samples(self) -> int:
```

Comment:
we should maybe be consistent throughout bofire. Either always `num_something` or `n_something` for number of something. I prefer the shorter one, of course ;).